providers/keycloak/src/airflow/providers/keycloak/auth_manager/routes/token.py

```
@@ -60,3 +64,23 @@ def create_token(body: TokenBody) -> TokenResponse:
     token = get_auth_manager().generate_jwt(user)
 
     return TokenResponse(access_token=token)
+
+
+@token_router.post(
+    "/refresh-token",
+    status_code=status.HTTP_200_OK,
+    responses=create_openapi_http_exception_doc([status.HTTP_400_BAD_REQUEST, status.HTTP_401_UNAUTHORIZED]),
+)
+def refresh_token(user: GetUserDep) -> TokenResponse:
```

`user: GetUserDep`
Should we use this dependency where it is from security, or should we obtain an access token as a request and return a token in response? I used it and it worked but not sure if we should call it from here, since it is an injection
cc: @pierrejeambrun @vincbeck

---

I first made it async and called `get_user_from_token` from the route, then converted it to this

---

I think that `GetUserDep` is validating claim and checking token validity. I think that the dep will raise an error if the token is actually expired, which is not what we want. (We still want to be able to hit that route with a valid refresh token and an expired token)

---

Unless I misunderstood we want the call to fail if the token expired? My understanding of the flow:
- User calls an API with an expired token (and with refresh token)
- API call fails
- Front-end catch this and kick off a refresh token flow using the refresh token

Do you have something else in mind?

---

@vincbeck Yes I agree with that. Here we are using the `GetUserDep` in the refresh token endpoint, which validates the token claims and validity I think, so basically we won't be able to actually refresh the token with an expired token and a valid refresh token. (Cause this GetUserDep will fail early). 

At least that's my concern.

---

@vincbeck Yes I agree with that. Here we are using the `GetUserDep` in the refresh token endpoint, which validates the token claims and validity I think, so basically we won't be able to actually refresh the token with an expired token and a valid refresh token. (Because this GetUserDep will fail early and we will not be able to call that refresh_token endpoint). 

At least that's my concern.

---

Oh wait, that makes me realize one thing too. Dependencies like this need a FastAPI context.

On the fab implementation of the refresh token we will most likely need a `Flask` application context. 

I think we need to do the same mechanism as we have for login and logout:
- public endpoint that will do a redirection to the appropriate auth manager endpoint.

---

Oh yes I see, sorry I did not look at the updated code. I agree with Pierre, we should follow the same logic as login etc. Here is what we should at a glance:

- No longer have `refresh_token` method in auth manager interface
- Add new method `get_url_refresh_token` which can return a `str` or `None`. If it is `None`, the auth manager does not have a refresh token flow (which is ok)
- When the front-end detects the token expires, it start a refresh token flow, which is basically redirecting to the `get_url_refresh_token` URL

As Pierre mentioned, we should definitely do it in multiple iteration and focusing on the back-end first is a good idea.

Thanks a lot Bugra for your patience and making so many iterations on that PR

---

Great info! Indeed, it worked in practice with postman, but I couldn't make any tests green while trying to add unit tests. 

Yeap, I got it. I thought we would still utilize the refresh method from the auth manager. Let me update and update the enpoint with redirect. I thought just acquiring the token with the new user info should be enough to update local storage in UI, which worked in the middleware, but it's good to know we need the context while doing the refresh and unifying solution in this case makes sense

All good, I am happier this way since the code will contain the best solution we come up with at the end and needs less and less update in the future. My authentication knowledge also enhanced a lot. Thanks for your patience and switft comments! :)